src/Nethermind/Nethermind.Consensus/Processing/ParallelProcessing/MultiVersionMemory.cs

```diff
@@ -0,0 +1,194 @@
+        .ToArray();
+
+    private readonly HashSet<TLocation>?[] _lastWrittenLocations = new HashSet<TLocation>?[txCount];
+    private readonly HashSet<Read<TLocation>>?[] _lastReads = new HashSet<Read<TLocation>>[txCount];
+
+    private void ApplyWriteSet(Version version, Dictionary<TLocation, byte[]> writeSet)
+    {
+        ConcurrentDictionary<TLocation, Value> txData = _data[version.TxIndex];
+        foreach (KeyValuePair<TLocation, byte[]> write in writeSet)
+        {
+            txData[write.Key] = new(version.Incarnation, write.Value);
+        }
+    }
+
+    private bool UpdateWrittenLocations(int txIndex, Dictionary<TLocation, byte[]> writeSet)
+    {
+        ConcurrentDictionary<TLocation, Value> txData = _data[txIndex];
+        ref HashSet<TLocation>? lastWritten = ref _lastWrittenLocations[txIndex];
+        lastWritten ??= new HashSet<TLocation>();
+
+        if (lastWritten.Count != 0)
+        {
+            using ArrayPoolList<TLocation> toRemove = new(lastWritten.Count);
+            foreach (var id in lastWritten)
+            {
+                if (!writeSet.ContainsKey(id))
+                {
+                    toRemove.Add(id);
+                }
+            }
+
+            foreach (var id in toRemove)
+            {
+                lastWritten.Remove(id);
+                txData.TryRemove(id, out _);
+            }
+        }
+
+        int oldCount = lastWritten.Count;
+        lastWritten.UnionWith(writeSet.Keys);
+        return lastWritten.Count > oldCount;
+    }
+
+    public bool Record(Version version, HashSet<Read<TLocation>> readSet, Dictionary<TLocation, byte[]> writeSet)
+    {
+        if (typeof(TLogger) == typeof(IsTracing)) parallelTrace.Add($"{version} Record read-set: {{{string.Join(",", readSet.Select(r => $"{r.Location}:{r.Version}"))}}}, write-set: {{{string.Join(",", writeSet.Select(r => $"{r.Key}:{r.Value.ToHexString()}"))}}}.");
+        ApplyWriteSet(version, writeSet);
+        bool wroteNewLocation = UpdateWrittenLocations(version.TxIndex, writeSet);
+        _lastReads[version.TxIndex] = readSet;
+        return wroteNewLocation;
+    }
+
+    public void ConvertWritesToEstimates(ushort txIndex)
+    {
+        if (typeof(TLogger) == typeof(IsTracing)) parallelTrace.Add($"Tx {txIndex} ConvertWritesToEstimates.");
+        ConcurrentDictionary<TLocation, Value> txData = _data[txIndex];
+        HashSet<TLocation>? previousLocations = _lastWrittenLocations[txIndex];
+        if (previousLocations is not null)
+        {
+            foreach (var location in previousLocations)
+            {
+                txData[location] = Estimate;
+            }
+        }
+    }
+
+    public Status TryRead(TLocation location, ushort txIndex, out Version version, out byte[]? value)
+    {
+        long id = parallelTrace.ReserveId();
+        ushort prevTx = txIndex;
+        prevTx--;
+        while (prevTx != ushort.MaxValue)
```

Comment:
A magic const.

Reply:
not really, it is about underflow (going below 0)